File: app/src/main/java/net/emojiparty/android/jishotomo/data/room/Sense.java
package net.emojiparty.android.jishotomo.data.room;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Index;
import androidx.room.PrimaryKey;
import androidx.annotation.NonNull;

@Entity(tableName = "senses", indices = {@Index("entry_id")})
public class Sense {
  @PrimaryKey
  @NonNull
  private int id;

  // ints are non-null by default
  @NonNull
  @ColumnInfo(name = "entry_id")
  private int entryId;

  @ColumnInfo(name = "parts_of_speech")
  private String partsOfSpeech;

  @NonNull
  private String glosses;

  @ColumnInfo(name = "applies_to")
  private String appliesTo;

  @ColumnInfo(name = "cross_references")
  private String crossReferences;

  @NonNull public int getId() {
    return id;
  }

  public void setId(@NonNull int id) {
    this.id = id;
  }

  public int getEntryId() {
    return entryId;
  }

  public void setEntryId(int entryId) {
    this.entryId = entryId;
  }

  public String getPartsOfSpeech() {
    return partsOfSpeech;
  }

  public void setPartsOfSpeech(String partsOfSpeech) {
    this.partsOfSpeech = partsOfSpeech;
  }

  @NonNull public String getGlosses() {
    return glosses;
  }

  public void setGlosses(@NonNull String glosses) {
    this.glosses = glosses;
  }

  public String getAppliesTo() {
    return appliesTo;
  }

  public void setAppliesTo(String appliesTo) {
    this.appliesTo = appliesTo;
  }

  public String getCrossReferences() {
    return crossReferences;
  }

  public void setCrossReferences(String crossReferences) {
    this.crossReferences = crossReferences;
  }
}

File: app/src/main/java/net/emojiparty/android/jishotomo/data/room/Sense.kt
package net.emojiparty.android.jishotomo.data.room

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.Index
import androidx.room.PrimaryKey
import net.emojiparty.android.jishotomo.data.models.CrossReferencedEntry

/**
 * A Sense is like a definition, but is also self-referential,
 * because it can also cross-reference other Senses belonging to different Entries.
 *
 * In its most basic form, the cross references are stored as a semi-colon-separated
 * string of expressions. This is not used in this app, but was used to create the join
 * table represented by CrossReference.
 */
@Entity(tableName = "senses", indices = [Index("entry_id")])
data class Sense(
  @PrimaryKey var id: Int,
  @ColumnInfo(name = "entry_id") var entryId: Int,
  @ColumnInfo(name = "parts_of_speech") var partsOfSpeech: String? = null,
  var glosses: String,
  @ColumnInfo(name = "applies_to") var appliesTo: String? = null,
  @ColumnInfo(name = "cross_references") var _crossReferences: String? = null
) {
  @Ignore var crossReferences: List<CrossReferencedEntry> = emptyList()
}
